callable/src/main/java/io/grpc/callable/Callable.java

```diff
@@ -0,0 +1,423 @@
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.callable;
+
+import com.google.common.base.Supplier;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ExperimentalApi;
+import io.grpc.MethodDescriptor;
+import io.grpc.StatusException;
+import io.grpc.stub.ClientCalls;
+import io.grpc.stub.StreamObserver;
+
+import java.util.Iterator;
+import java.util.concurrent.Executor;
+
+import javax.annotation.Nullable;
+
+/**
+ * A callable is an object which represents one or more rpc calls. Various operators on callables
+ * produce new callables, representing common API programming patterns. Callables can be used to
+ * directly operate against an api, or to efficiently implement wrappers for apis which add
+ * additional functionality and processing.
+ *
+ * <p>Technically, callables are a factory for grpc {@link ClientCall} objects, and can be executed
+ * by methods of the {@link ClientCalls} class. They also provide shortcuts for direct execution of
+ * the callable instance.
+ */
+@ExperimentalApi
+public abstract class Callable<RequestT, ResponseT> {
+
+  // TODO(wrwg): Support interceptors and method option configurations.
+  // TODO(wrwg): gather more feedback whether the overload with java.util.Concurrent hurts that
+  //             much that we want to rename this into ClientCallable or such.
+
+  // Subclass Contract
+  // =================
+
+  /**
+   * Creates a new GRPC call from this callable. A channel may or may not be provided.
+   * If a channel is not provided, the callable must be bound to a channel.
+   */
+  public abstract ClientCall<RequestT, ResponseT> newCall(@Nullable Channel channel);
```

Comment:
The canonical use case is this one:

    class MyApi {
      private final static Callable<I, O> GET_FEATURE = 
          Callable.create(MyGrpc.Descriptors.GET_FEATURE).retrying();

      private final Channnel channel;
      MyApi(Channel channel) {
          this.channel = channel;
      }
      ...
      public Callable<I, O> getFeatureCall() {
        return GET_FEATURE.bind(channel);
      }
    }

I think we should keep that feature for now (note Callable is marked as experimental), and then discuss if we have more usage whether we remove either pre-bound or unbound, or keep both.